### PR TITLE
issue:vulnera#32

### PR DESCRIPTION
**Pull request by AntonioliBenjamin**

### CODE REVIEW

**Comment by AntonioliBenjamin** [comment]
I can't find a way to test that there are two requests sent, the "fetchDataForPackageURLs" function seems to be OK

File: src/strategies/sonatype.js [hunk 1]
```diff
@@ -51,18 +51,44 @@ function createPackageURLCoordinates([dependencyName, dependencyPayload]) {
...
+
+  for (let i = 0; i < arr.length; i += chunkSize) {
+    chunkedArr.push(arr.slice(i, i + chunkSize));
+  }
+
+  return chunkedArr;
+}
+
+export async function fetchDataForPackageURLs(coordinates) {
+  const chunkSize = 128;
```
Comment (fraxken) [inline src/strategies/sonatype.js:65]: You can declare this variable as a CONSTANT at the top of the code

```js
// CONSTANTS
const kChunkSize = 128;
```

File: src/strategies/sonatype.js [hunk 2]
```diff
@@ -51,18 +51,44 @@ function createPackageURLCoordinates([dependencyName, dependencyPayload]) {
...
   };
 
   try {
-    const { data } = await httpie.post(kSonatypeApiURL, requestOptions);
-
-    return data;
+    if(coordinates.length < chunkSize) {
+      const { data } = await httpie.post(kSonatypeApiURL, requestOptions);
+      return data 
+    }
```
Comment (fraxken) [inline src/strategies/sonatype.js:78]: If `chunkArray` work as expected, I don't think we need to handle the case where there is only one request

File: src/strategies/sonatype.js [hunk 3]
```diff
@@ -51,18 +51,44 @@ function createPackageURLCoordinates([dependencyName, dependencyPayload]) {
...
+    const { data } = chunks.map(chunk => {
+      return httpie.post(kSonatypeApiURL, {
+        headers: {
+          accept: "application/json"
+        },
+        body: { coordinates: chunk }
+      });
+    })
+   
+    return await Promise.all(data)
```
Comment (fraxken) [inline src/strategies/sonatype.js:91]: There is an error/problem here for sure. First this is not possible to use a destructuration on a Array of promises
```js
const { data } = chunks.map
```

And `return await Promise.all(data)` is not the same as `return data` at the top.

File: src/strategies/sonatype.js [hunk 4]
```diff
@@ -51,18 +51,44 @@ function createPackageURLCoordinates([dependencyName, dependencyPayload]) {
...
-async function fetchDataForPackageURLs(coordinates) {
+export function chunkArray(arr, chunkSize) {
+  let chunkedArr = [];
+
+  for (let i = 0; i < arr.length; i += chunkSize) {
+    chunkedArr.push(arr.slice(i, i + chunkSize));
+  }
+
+  return chunkedArr;
+}
```
Comment (fraxken) [inline src/strategies/sonatype.js:62]: You can use a Generator function here

```js
export function* chunkArray(arr) {
  for (let i = 0; i < arr.length; i += kChunkSize) {
    yield arr.slice(i, i + kChunkSize);
  }
}
```

**Review by fraxken — CHANGES_REQUESTED** [review]

File: test/strategies/sonatype/index.unit.js [hunk 5]
```diff
@@ -132,3 +133,30 @@ test("sonatype strategy: hydratePayloadDependencies when using NodeSecure standa
   restoreHttpAgent();
   tape.end();
 });
+
+
+test("sonatype strategy: fetchDataForPackageURLs with chunked coordinates", async(tape) => {
+  const [mockedHttpAgent, restoreHttpAgent] = setupHttpAgentMock();
+  const mockedHttpClient = mockedHttpAgent.get(kSonatypeOrigin);
+
+  let chunkedCoordinates = []
```
Comment (antoine-coulon) [inline test/strategies/sonatype/index.unit.js:142]: I would not call that `chunkedCoordinates` here as they are not chunked yet, also you could do that to generate your collection of coordinates:

```js
const coordinates = Array.from({ length: 200 }, () => kFakePackageURL);
```

**Review by antoine-coulon — COMMENTED** [review]
Not much to add from what @fraxken suggested :) 

Be sure to use the ESLint from the repository configuration, it seems that in the `test/strategies/sonatype/index.unit.js` format is not working as semicolons are some missing

File: test/strategies/sonatype/index.unit.js (hunk 5 above)
Comment (AntonioliBenjamin) [reply test/strategies/sonatype/index.unit.js:142]: done =)

File: src/strategies/sonatype.js (hunk 4 above)
Comment (AntonioliBenjamin) [reply src/strategies/sonatype.js:62]: done !

File: src/strategies/sonatype.js (hunk 1 above)
Comment (AntonioliBenjamin) [reply src/strategies/sonatype.js:65]: done

**Comment by fraxken** [comment]
@AntonioliBenjamin I think you forgot to push ? (because there is no updates)

**Comment by AntonioliBenjamin** [comment]
No, I haven't finished all the changes yet.

**Pull request closed without merging**